### PR TITLE
PHOENIX-5786 Update phoenix-hive to support current hive versions

### PR DESCRIPTION
**Pull request by richardantal**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -60,6 +61,8 @@
     <test.output.tofile>true</test.output.tofile>
     <top.dir>${project.basedir}</top.dir>
 
+    <jdk.version>1.7</jdk.version>
```
Comment (joshelser) [inline pom.xml:64]: Do we need to be supporting Java7 for anything? Can we move to 8?

**Comment by joshelser** [comment]
Running a `mvn verify` on HEAD of this pull request, fails on the phoenix-hive module with a Jersey compatibility issue:
```
2020-07-17 12:59:35,108 ERROR [main] org.mortbay.log: Error starting handlers
java.lang.IncompatibleClassChangeError: com.sun.jersey.json.impl.provider.entity.JSONRootElementProvider and com.sun.jersey.json.impl.provider.entity.JSONRootElementProvider$Wadl disagree on InnerClasses attribute
	at java.lang.Class.getDeclaringClass0(Native Method)
	at java.lang.Class.getDeclaringClass(Class.java:1235)
	at java.lang.Class.getEnclosingClass(Class.java:1277)
	at com.sun.jersey.core.spi.component.ComponentConstructor.getInstance(ComponentConstructor.java:169)
	at com.sun.jersey.core.spi.component.ProviderFactory.__getComponentProvider(ProviderFactory.java:166)
	at com.sun.jersey.core.spi.component.ProviderFactory.getComponentProvider(ProviderFactory.java:137)
	at com.sun.jersey.core.spi.component.ProviderServices.getComponent(ProviderServices.java:256)
	at com.sun.jersey.core.spi.component.ProviderServices.getServices(ProviderServices.java:160)
	at com.sun.jersey.core.spi.factory.MessageBodyFactory.initReaders(MessageBodyFactory.java:176)
	at com.sun.jersey.core.spi.factory.MessageBodyFactory.init(MessageBodyFactory.java:162)
	at com.sun.jersey.server.impl.application.WebApplicationImpl._initiate(WebApplicationImpl.java:1277)
	at com.sun.jersey.server.impl.application.WebApplicationImpl.access$700(WebApplicationImpl.java:169)
	at com.sun.jersey.server.impl.application.WebApplicationImpl$13.f(WebApplicationImpl.java:775)
	at com.sun.jersey.server.impl.application.WebApplicationImpl$13.f(WebApplicationImpl.java:771)
	at com.sun.jersey.spi.inject.Errors.processWithErrors(Errors.java:193)
	at com.sun.jersey.server.impl.application.WebApplicationImpl.initiate(WebApplicationImpl.java:771)
	at com.sun.jersey.guice.spi.container.servlet.GuiceContainer.initiate(GuiceContainer.java:121)
	at com.sun.jersey.spi.container.servlet.ServletContainer$InternalWebComponent.initiate(ServletContainer.java:318)
	at com.sun.jersey.spi.container.servlet.WebComponent.load(WebComponent.java:609)
	at com.sun.jersey.spi.container.servlet.WebComponent.init(WebComponent.java:210)
	at com.sun.jersey.spi.container.servlet.ServletContainer.init(ServletContainer.java:373)
	at com.sun.jersey.spi.container.servlet.ServletContainer.init(ServletContainer.java:710)
	at com.google.inject.servlet.FilterDefinition.init(FilterDefinition.java:114)
	at com.google.inject.servlet.ManagedFilterPipeline.initPipeline(ManagedFilterPipeline.java:98)
	at com.google.inject.servlet.GuiceFilter.init(GuiceFilter.java:172)
	at org.mortbay.jetty.servlet.FilterHolder.doStart(FilterHolder.java:97)
	at org.mortbay.component.AbstractLifeCycle.start(AbstractLifeCycle.java:50)
	at org.mortbay.jetty.servlet.ServletHandler.initialize(ServletHandler.java:713)
	at org.mortbay.jetty.servlet.Context.startContext(Context.java:140)
	at org.mortbay.jetty.webapp.WebAppContext.startContext(WebAppContext.java:1282)
	at org.mortbay.jetty.handler.ContextHandler.doStart(ContextHandler.java:518)
	at org.mortbay.jetty.webapp.WebAppContext.doStart(WebAppContext.java:499)
	at org.mortbay.component.AbstractLifeCycle.start(AbstractLifeCycle.java:50)
	at org.mortbay.jetty.handler.HandlerCollection.doStart(HandlerCollection.java:152)
	at org.mortbay.jetty.handler.ContextHandlerCollection.doStart(ContextHandlerCollection.java:156)
	at org.mortbay.component.AbstractLifeCycle.start(AbstractLifeCycle.java:50)
	at org.mortbay.jetty.handler.HandlerWrapper.doStart(HandlerWrapper.java:130)
	at org.mortbay.jetty.Server.doStart(Server.java:224)
	at org.mortbay.component.AbstractLifeCycle.start(AbstractLifeCycle.java:50)
	at org.apache.hadoop.http.HttpServer2.start(HttpServer2.java:877)
	at org.apache.hadoop.yarn.webapp.WebApps$Builder.start(WebApps.java:306)
	at org.apache.hadoop.mapreduce.v2.app.client.MRClientService.serviceStart(MRClientService.java:144)
	at org.apache.hadoop.service.AbstractService.start(AbstractService.java:193)
	at org.apache.hadoop.mapreduce.v2.app.MRAppMaster.serviceStart(MRAppMaster.java:1147)
	at org.apache.hadoop.service.AbstractService.start(AbstractService.java:193)
	at org.apache.hadoop.mapreduce.v2.app.MRAppMaster$5.run(MRAppMaster.java:1552)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:422)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1754)
	at org.apache.hadoop.mapreduce.v2.app.MRAppMaster.initAndStartAppMaster(MRAppMaster.java:1548)
	at org.apache.hadoop.mapreduce.v2.app.MRAppMaster.main(MRAppMaster.java:1481)
2020-07-17 12:59:35,113 INFO [main] org.mortbay.log: Started HttpServer2$SelectChannelConnectorWithSafeStartup@0.0.0.0:60485
2020-07-17 12:59:35,113 ERROR [main] org.apache.hadoop.mapreduce.v2.app.client.MRClientService: Webapps failed to start. Ignoring for now:
org.apache.hadoop.yarn.webapp.WebAppException: Error starting http server
	at org.apache.hadoop.yarn.webapp.WebApps$Builder.start(WebApps.java:310)
	at org.apache.hadoop.mapreduce.v2.app.client.MRClientService.serviceStart(MRClientService.java:144)
	at org.apache.hadoop.service.AbstractService.start(AbstractService.java:193)
	at org.apache.hadoop.mapreduce.v2.app.MRAppMaster.serviceStart(MRAppMaster.java:1147)
	at org.apache.hadoop.service.AbstractService.start(AbstractService.java:193)
	at org.apache.hadoop.mapreduce.v2.app.MRAppMaster$5.run(MRAppMaster.java:1552)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:422)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1754)
	at org.apache.hadoop.mapreduce.v2.app.MRAppMaster.initAndStartAppMaster(MRAppMaster.java:1548)
	at org.apache.hadoop.mapreduce.v2.app.MRAppMaster.main(MRAppMaster.java:1481)
Caused by: java.io.IOException: Problem in starting http server. Server handlers failed
	at org.apache.hadoop.http.HttpServer2.start(HttpServer2.java:889)
	at org.apache.hadoop.yarn.webapp.WebApps$Builder.start(WebApps.java:306)
	... 10 more
```

Maybe this being broken is out of scope for this issue

File: phoenix-hive3/pom.xml [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+  </parent>
+  <artifactId>phoenix-hive3</artifactId>
+  <name>Phoenix - Hive3</name>
+<properties>
+    <test.tmp.dir>${project.build.directory}/tmp</test.tmp.dir>
+    <netty.version>4.1.47.Final</netty.version>
+    <phoenix.version>5.1.0-SNAPSHOT</phoenix.version>
+    <hbase.version>2.2.1</hbase.version>
+    <hadoop.version>3.0.0</hadoop.version>
+    <hadoop-two.version>3.0.0</hadoop-two.version>
```
Comment (joshelser) [inline phoenix-hive3/pom.xml:40]: unused

File: phoenix-hive3/pom.xml [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+    <version>1.0.0-SNAPSHOT</version>
+  </parent>
+  <artifactId>phoenix-hive3</artifactId>
+  <name>Phoenix - Hive3</name>
+<properties>
+    <test.tmp.dir>${project.build.directory}/tmp</test.tmp.dir>
+    <netty.version>4.1.47.Final</netty.version>
+    <phoenix.version>5.1.0-SNAPSHOT</phoenix.version>
+    <hbase.version>2.2.1</hbase.version>
+    <hadoop.version>3.0.0</hadoop.version>
```
Comment (joshelser) [inline phoenix-hive3/pom.xml:39]: phoenix is at 3.0.3. Should we use that too?

File: phoenix-hive3/pom.xml [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+    <artifactId>phoenix-connectors</artifactId>
+    <version>1.0.0-SNAPSHOT</version>
+  </parent>
+  <artifactId>phoenix-hive3</artifactId>
+  <name>Phoenix - Hive3</name>
+<properties>
+    <test.tmp.dir>${project.build.directory}/tmp</test.tmp.dir>
+    <netty.version>4.1.47.Final</netty.version>
+    <phoenix.version>5.1.0-SNAPSHOT</phoenix.version>
+    <hbase.version>2.2.1</hbase.version>
```
Comment (joshelser) [inline phoenix-hive3/pom.xml:38]: And 2.2.4 for HBase from Phoenix

**Comment by joshelser** [comment]
This is what I have done:
* `mvn install` from master on phoenix.git
* `mvn verify -am -pl phoenix-hive3`

I then see test failures:
```
[INFO] Running org.apache.phoenix.hive.HiveMapReduceIT
[ERROR] Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 0.005 s <<< FAILURE! - in org.apache.phoenix.hive.HiveMapReduceIT
[ERROR] org.apache.phoenix.hive.HiveMapReduceIT  Time elapsed: 0.005 s  <<< ERROR!
java.lang.NoClassDefFoundError: org/apache/hadoop/crypto/key/KeyProviderTokenIssuer
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:756)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:468)
	at java.net.URLClassLoader.access$100(URLClassLoader.java:74)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:369)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:363)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:362)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:355)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
	at org.apache.phoenix.query.BaseTest.initMiniCluster(BaseTest.java:544)
	at org.apache.phoenix.query.BaseTest.setUpTestCluster(BaseTest.java:448)
	at org.apache.phoenix.query.BaseTest.checkClusterInitialized(BaseTest.java:434)
	at org.apache.phoenix.query.BaseTest.setUpTestDriver(BaseTest.java:520)
	at org.apache.phoenix.query.BaseTest.setUpTestDriver(BaseTest.java:515)
	at org.apache.phoenix.end2end.BaseHBaseManagedTimeIT.doSetup(BaseHBaseManagedTimeIT.java:57)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:24)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:55)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:137)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeLazy(JUnitCoreWrapper.java:119)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:87)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)
	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:158)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
Caused by: java.lang.ClassNotFoundException: org.apache.hadoop.crypto.key.KeyProviderTokenIssuer
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:355)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
	... 48 more
```
Seems like we're missing hadoop-common on the class path?

File: phoenix-hive3/pom.xml (hunk 4 above)
Comment (stoty) [reply phoenix-hive3/pom.xml:38]: At the moment Phoenix is set to use 2.2.4 by default, but I think that we should consider changing the default to the 2.1 profile (always the oldest supported), and publish the maven artifacts built with that. (Going to open a ticket for that right away)

For the same reason, building with 2.1.9 instead may be a better solution here.

File: pom.xml (hunk 1 above)
Comment (richardantal) [reply pom.xml:64]: There are significant changes in the java code between phoenix-hive and phoenix-hive, it is not just a copy and pom.xml change. 
Hive 3.x supports Hadoop 3 but hive 2 does not, Hive3 is intended to be used with Phoenix 5 meanwhile the old phoenix-hive connector is working with phoenix 4.

This JDK version is in connectors pom.xml, all the modules but the phoenix-hive3 uses jdk 1.7. In phoenix-hive3 it is being overwritten to 1.8

File: phoenix-hive3/pom.xml (hunk 4 above)
Comment (stoty) [reply phoenix-hive3/pom.xml]: But since we'll need to make another pass on the dependencies anyway, 2.2.4 is fine for now.

**Comment by stoty** [comment]
In general, I think that this needs a thorough review of the dependency handling. We are dependencyManaging some versions that we shouldn't, and there are dependencies and exclusions that I think are not needed.

However, fixing those would be much easier once we have something that we can use as a baseline, so I think that we should first commit this, and re-think the dependencies after.

+1 from me, but please get @joshelser 's approval too.

**Comment by stoty** [comment]
> As I understand it, this is largely just copy-paste'ing from `phoenix-hive` into a `phoenix-hive3`, and then modifying the pom.xml to pull in the correct dependencies. While this approach is effective, I'm worried about the long-term maintenance of supporting multiple versions of Hive via separate copies of code.
> 
> Maybe we can stop caring about Hive2 soon and we just drop the `phoenix-hive` module entirely?

The current Apache connector is actually based on Hive 1. I don't think it'd even work with Hive 2. This patch contains significant code changes to support the V3 API instead of the V1 one.

According to @richardantal's research Hive1 and Hive2 is only supported with Hadoop2, while Hive3+ is Hadoop3 only.
This in turn means that 4.x can only hope to support Hive1,2, while 5.x can only support Hive 3+.

Replacing phoenix-hive would mean that we'd drop hive connector support for 4.x completely. I don't know how large the Hive1 installed base is, and whether it'd be even possible for them to use a recent-ish Phoenix on such an old cluster, so it could be a good thing.

While there are significant changes between between the Hive 1 and Hive 2 APIs, the Hive2 and Hive3 APIs seem similar enough that it should be possible to backport this Hive3 connector for Hive2/Phoenix 4.x, and then refactor the Hive3 connector as an extension on top of the new Hive2 connector, though that would either need copying and overwriting the source files, or putting the hive2 and hive3 connectors into different java packages.

However, the Hive 2.x and 3.x connectors would need to live in their own their own Hive version specific sub-projects, so I still think that putting this into a new repo is a good idea.

**Comment by stoty** [comment]
Maybe pre-emptively move this code to org.apache.phoenix.hive3 package now, so that we don't have to deal with the 2.x vs 3.x conflicts later ?

File: phoenix-hive3/pom.xml (hunk 4 above)
Comment (joshelser) [reply phoenix-hive3/pom.xml]: 2.1.9 or 2.2.4 are fine by me, either way :)

**Comment by joshelser** [comment]
> Maybe pre-emptively move this code to org.apache.phoenix.hive3 package now, so that we don't have to deal with the 2.x vs 3.x conflicts later ?

That's a really smart idea. Hive's classpath gets the entire world, most of the time.

File: pom.xml (hunk 1 above)
Comment (joshelser) [reply pom.xml:64]: Ok, cool. I didn't have an easy way to validate this. Copying is probably easiest! Thanks!

**Comment by joshelser** [comment]
Rerunning tests worked this time. Not sure why they failed the first time.

> I think that this needs a thorough review of the dependency handling. We are dependencyManaging some versions that we shouldn't, and there are dependencies and exclusions that I think are not needed.

Do you think we should do this now, @stoty?  I am OK to defer dependency cleanup to having a "good" state of this branch with hive3, and then try to make incremental changes to reduce subsequent "dependency cleanup" patches.

You have my +1 if you want to merge this one now, Istvan, but I respect the decision to clean up dependencies right away.

**Comment by stoty** [comment]
> > Maybe pre-emptively move this code to org.apache.phoenix.hive3 package now, so that we don't have to deal with the 2.x vs 3.x conflicts later ?
> 
> That's a really smart idea. Hive's classpath gets the entire world, most of the time.

I've slept on (literally), and now I think that that's not the way to, as this package name is exposed to the Hive user-facing API, and having to deal with version specific package names would be an unnecessary inconvenience for the users.

We should just have have something like phoenix-hbase-common, where we make the version-dependent classes abstract, and build the Hive2 and Hive3 connectors on top of it. 

This is of course just a plan for the future connector for 4.x, not for the present patch.

> You have my +1 if you want to merge this one now, Istvan, but I respect the decision to clean up dependencies right away.

Thank you. I'd like to get the code in ASAP, and deal with the dependencies and possible shaded JAR later.

**Comment by stoty** [comment]
I suggest that try adding the JVM log files as artifacts to the Github workflow:
https://docs.github.com/en/actions/configuring-and-managing-workflows/persisting-workflow-data-using-artifacts

**Comment by stoty** [comment]
You could also look into why those tests are taking so long. I suspect that that they can be made faster without losing functionality.

**Comment by stoty** [comment]
Some minor issues:

- Hive-exec dependency is duplicated
- Pom indentation (since it's technically a new POM, you may want to have your IDE reformat it)
- You may want to rename the hive version property to hive3.version, as it is distinct from the other hive connector's hive version
- Netty dependency is in the test section, but not marked as test scope (probably just add test scope)

**Pull request closed without merging**